src/main/java/org/opentripplanner/middleware/docs/PublicApiDocGenerator.java

```diff
@@ -0,0 +1,416 @@
+            }
+        }
+        return false;
+    }
+    /**
+     * Determines whether a given path is restricted (non-public).
+     */
+    private boolean isRestricted(String pathName) {
+        return !isPublic(pathName);
+    }
+
+    /**
+     * Edit array types (e.g. MonitoredTrip[]) found in the definitions sections.
+     * (These are generated by us, using an array as return type, bypassing a spark-swagger limitation.)
+     */
+    private void inlineArrayDefinitions() {
+        final String ARRAY_MARKER = "[]";
+
+        // Find return type references from all "schema" entries.
+        for (JsonNode node: rootNode.findValues("schema")) {
+            ObjectNode schemaNode = (ObjectNode) node;
+            JsonNode refNode = node.get(REF_FIELD);
+            if (refNode != null && refNode.asText().endsWith(ARRAY_MARKER)) {
+                String baseType = refNode.asText().substring(DEFINITIONS_REF.length()).replace(ARRAY_MARKER, "");
+
+                // Replace the ref node with the inlined array definition, so the schema node becomes:
+                //   schema:
+                //     type: array
+                //     items:
+                //       $ref: "#/definitions/MonitoredTrip"
+                schemaNode.remove(REF_FIELD);
+                schemaNode
+                    .put("type", "array")
+                    .set("items", JsonNodeFactory.instance.objectNode()
+                        .put(REF_FIELD, DEFINITIONS_REF + baseType));
+
+                // Delete the autogenerated array definition.
+                definitionsNode.remove(baseType + ARRAY_MARKER);
+            }
+        }
+    }
+
+    /**
+     * Generate missing types not created by spark-swagger.
+     * Note: This method might not end up creating types if there are no missing types.
+     */
+    private void generateMissingTypes() {
+        // Find the types from all $ref entries.
+        for (JsonNode node: rootNode.findValues(REF_FIELD)) {
+            if (node.asText().startsWith(DEFINITIONS_REF)) {
+                String baseType = node.asText().substring(DEFINITIONS_REF.length());
+
+                // Add type if it is not defined.
+                if (definitionsNode.get(baseType) == null) {
+                    JsonNode typeDefinition = JsonNodeFactory.instance.objectNode()
+                        .put("type", "object")
+                        .put("description", "autogenerated");
+
+                    definitionsNode.set(baseType, typeDefinition);
+                }
+            }
+        }
+    }
+
+    /**
+     * Removes unused types (not referenced in a $ref entry anywhere) from the definitions section.
+     */
+    private void removeUnusedTypes() {
+        int unusedTypeCount;
+        // Execute until there are no unused types.
+        do {
+            // Find return type references anywhere (only store one instance of each).
+            LinkedHashSet<String> typeReferences = new LinkedHashSet<>();
```

Comment:
I learned a new thing about what a LinkedHashSet is, but do we really need it here? Wouldn't a regular HashSet suffice?